오픈 마켓 [STEP1] 예거, 릴리

yeahg-dev opened the pull request:
안녕하세요 찰리 @kcharliek 
릴리와 예거입니다!
STEP1-1과 STEP1-2의 네트워크 담당 타입 설계까지 완료하여 DraftPR올립니다.
찰리가 강의해주신 부분을 최대한 녹여보려고 했으나 아직 부족한 점이 많습니다🥲
바쁘심에도 항상 신경써주셔서 정말 감사합니다 🙇🏻‍♀️ 🙇 
PR 잘 부탁드립니다!
<br>
 
# 🛠️ 구현 내용
## STEP 1-1
- API 통신 Request와 Response에 사용 될 JSON객체를 먼저 사용자 정의 타입으로 구현했습니다.
  - APIRequest에서만 사용되는 타입은 `Network` > `RequestType`에 파일 분류하여 찾기 쉽도록 했습니다.
  - APIResponse에서 사용되는 타입은 앱의 Model의 부분이라 판단해서 `Model` 폴더에 분류했습니다.
- `MockJasonParser`타입을 구현해서 네트워크 통신이 불가능한 상황에서 mockResponse객체가 파싱이 정상적으로 되는지 테스트했습니다.

## STEP 1-2
- `APIURL`열거형의 연산프로퍼티로 각 API에 대한 URL을 만들도록 구현했습니다.
- `APIRequestProtocol` 기본 구현을 활용했습니다.
  - APIRequest를 호출할 때 공통적으로 쓰이는 메서드와 프로퍼티를 기본구현으로 만들어줘서, 
  - `APIRequestProtocol`을 채택하는 타입에서 사용할 수 있도록 했습니다.
 - `APIRequestProtocol`이 Parserable`프로토콜을 상속받도록 했습니다.
    - JSON객체를 encode, decode하는 메서드를 기본 구현했습니다.
    - 확장성을 고려해서 `Utility`폴더에 분류해주었습니다. 
  - 실제 API를 호출할 곳에서 쓰일 메서드를 `APIRequest`열거형의 static method로 구현했습니다. 
   - 열거형의 static method를 각각의 struct로 분리해주는 방향으로 리팩터링해볼 계획입니다.
<br>

# 🔥  고민한 내용

### 1️⃣ dateDecodingStrategy.iso8601 사용 실패

response 로 들어오는 날짜(Date) 포맷이 `iso8601` 포맷과 유사하다는 걸 포착했습니다.
> 예시) 2022-01-04T00:00:00.00

<p align="left"><img src="https://user-images.githubusercontent.com/71127966/148515308-64bbc2af-550a-4465-a0e1-ac116ebf03fb.png" width="45%"></p> 

마음에 걸리는 건, 마지막 milliseconds 의 자리수가 하나 다르다는 것이었습니다.
iso8601 에서는 `3-digits`인데, 저희 response 는 `2-digit` 이었거든요.

실험을 위해 `iso8601 전략`을 채택하고 Mock JSON 데이터에 대해 유닛 테스트를 작성해 돌려보니 디코딩에 실패했습니다.
[제드 블로그](https://zeddios.tistory.com/397)에서 `"위 그림의 포맷이어야만 인식을 합니다."` 라는 내용이 있어서, 저희는 iso8601 포맷을 포기하고 `커스텀 포맷`을 선택했습니다.

```swift
let dateFormatter = DateFormatter()
dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SS"
decoder.dateDecodingStrategy = .formatted(dateFormatter)
```

<br>

### 2️⃣ convertFromSnakeCase 전략 🆚 CodingKeys 열거형

keyDecodingStrategy 중의 `convertFromSnakeCase` 전략과 `CodingKeys 열거형` 중에 어떤 것을 사용할 것인지 논의했습니다.
찾아보니까 디코딩할 때는 `Strategy 가 먼저 적용`된 이후에 2순위로 CodingKeys 가 적용된다는 내용을 확인했습니다. [출처](https://stackoverflow.com/questions/49881621/the-convertfromsnakecase-strategy-doesnt-work-with-custom-codingkeys-in-swi)

저희는 [스타일쉐어 Swift Style Guide #약어](https://github.com/StyleShare/swift-style-guide)의 기준대로 네이밍을 하고 있었는데요.
`thumbnailURL` 같은 상수 네이밍에서, `URL` 을 대문자로 선언한 뒤에, convertFromSnakeCase 전략을 사용하면, `Url` 로 들어오게 되니까 디코딩이 실패하는 결과로 이어졌습니다.
따라서, 네이밍에서의 충돌을 피하려면 둘 중 하나만 사용해야 했기에, 저희는 장단점을 비교해봤습니다.

- **convertFromSnakeCase 전략**
  - 🟢  장점 -> CodingKeys 를 매번 선언해줄 필요가 없어서, 모델의 코드가 간결해짐
  - 🔴  단점 -> 타입의 모든 프로퍼티가 반드시 lowerCamelCase 로 작성되어야 함

- **CodingKeys 열거형**
  - 🟢  장점 -> 커스터마이징이 수월해져서, 원하는 네이밍을 만들어낼 수 있음
  - 🔴  단점 -> 타입 밑에 매번 선언해줘야 해서 불편하고, 코드의 양이 많아짐

저희 둘 다 `CodingKeys` 를 쓰는 게 ‘확장성’이나 ‘유연성’에서 더 장점이 있다고 의견을 모았습니다.
참고로 [convertFromSnakeCase 공식문서](https://developer.apple.com/documentation/foundation/jsondecoder/keydecodingstrategy/convertfromsnakecase?changes=latest_minor)에 `“Acronyms(두문자어)”`은 제대로 처리하기 어렵다는 노트가 써있기도 합니다. 그런 경우에는 CodingKeys 열거형을 사용하라고 적혀있더라구요. 😄 

<p align="center"><img src="https://user-images.githubusercontent.com/71127966/148519336-8716d9db-ee26-4e52-a46d-2161768af0ec.png" width="80%"></p> 

<br>

### 3️⃣ Request에서 boundary를 꼭 써야할까?
찰리에게 DM으로 여쭤본 내용인데요~
서버 API문서의 Example Request의 형식에는 multipart form data이지만 boundary를 설정해주지 않는 것을 확인했습니다.
대신 `--form`이란 걸 사용해서 구분해줬나 추측해보았습니다🙄
![image](https://user-images.githubusercontent.com/81469717/148518140-a5f35980-6551-4b45-b6e4-0d5e1efc2575.png)

[모질라 문서](https://developer.mozilla.org/ko/docs/Web/HTTP/Headers/Content-Type)에서는 multi part form 데이터는 boundary로 구분을 필수적으로 해줘야한다는 내용이 있어서 고민이 되었습니다🤔
<img width="786" alt="스크린샷 2022-01-07 오후 5 54 18" src="https://user-images.githubusercontent.com/81469717/148518455-1459e04c-037b-4c2f-b5f5-ec0947d58202.png">

이번 기회를 통해 조금 더 공부해보고자 boundary를 적용시켜보았습니다.

`ContentDispsition~` 과 `--boundary(UUID사용)`을 body에 들어갈 content사이 사이에 추가해봤습니다.
그리고 `Content-Disposition: form-data; name=\(imageFile.key)`의 name에 parameter로 Request에서 보낼 name을 주입받도록 구현해보았습니다.
multipart form data를 바디에 넣어줄 때 
part별 name을 Content-Disposition에 넣어주고, data를 아래에 추가해주는게 맞는 방법일까요?
```swift
static func createBody(params productInfo: NewProductInfo, images: [ImageFile]) -> Data? {
        var body = Data()
        let lineBreak = "\r\n"
        
        let params = "Content-Disposition: form-data; name=\"params\"\(lineBreak)"
        guard let encodedProductInfo = encode(from: productInfo) else {
            return nil
        }
        body.appendString(params)
        body.append(encodedProductInfo)
        body.appendString("\(lineBreak)\(lineBreak)")

        images.forEach { imageFile in
            body.appendString("Content-Disposition: form-data; name=\(imageFile.key); filename=\(imageFile.fileName)\(lineBreak)")
            body.appendString("Content-Type: \(imageFile.type.description)\(lineBreak)")
            body.appendString("\(imageFile.data)\(lineBreak)")
        }
        
        body.appendString("\(boundary)--")
        return body
    }
```

<br>

# 🤔 궁금한 점
### 1️⃣ 어떻게 추상화 레벨을 높일 수 있을까요?
 찰리가 저희에게 설명해주신 대로, 추상화 레벨을 높게 설계해보려 했으나, 처음부터 복잡한 구조로 설계하려니까 난이도가 너무 높았습니다. 😭 
그래서 구현을 먼저 하고, 그 뒤에 리팩토링을 통해 추상화를 시도해보려고 하는데요.

먼저, 공통적으로 사용되는 메서드에 대해 프로토콜 기본구현 형태로 분리시켰습니다.
현재 저희 코드 구조에서 추상화 레벨을 높이려면, 어디를 어떻게 바꿔야 할지 조언을 해주실 수 있을까요?

### 2️⃣ boundary는 request마다 생성해주어야하는 건지, 재사용이 가능한건지 궁금합니다!

<br>

긴 PR 읽어주셔서 감사합니다🙏🏻

Jager-yoo committed:
file: mock JSON 파일 추가

yeahg-dev committed:
file: 파일 역할에 따른 디렉토리 분류

Jager-yoo committed:
file: mock JSON 파일 업데이트, 디렉토리 순서 변경

yeahg-dev committed:
feat: Product 타입 구현

Jager-yoo committed:
feat: ProductsListPage 타입 구현

yeahg-dev committed:
feat: Currency 타입 구현

- Product 타입의 currencyCode 연산프로퍼티 구현

Jager-yoo committed:
MockJSONParser 구현, 테스트 케이스 초기 설정

Jager-yoo committed:
test: MockJSONParser 구현, 테스트 케이스 작성

- Product 타입의 createdAt, issuedAt 상수 타입을 String -> Date 변경
- products.json 내부 줄바꿈 수정

yeahg-dev committed:
Image, Vendor타입 구현, keyDecodingStrategy대신 CodingKeys 선택

Jager-yoo committed:
refactor: 모델 타입에 CodingKeys 열거형 구현

- currency 관련 연산 프로퍼티 삭제

yeahg-dev committed:
feat: ProductDetail 타입 구현

yeahg-dev committed:
feat: HTTPMethod타입, APIRequest프로토콜 구현

Jager-yoo committed:
feat: HealthCheckerRequest 타입 구현

- APIRequestInfoOwner 프로토콜 기본구현
- 모델 타입들이 APIResponse 프로토콜 채택

yeahg-dev committed:
feat: APIURL 타입 구현

Jager-yoo committed:
feat: execute, requestHealthChecker 메서드 구현

yeahg-dev committed:
feat: Request관련 ProductInfo 타입 2개 구현

Jager-yoo committed:
feat: Parser 타입 구현, requestProductRegistration 중간 저장

Jager-yoo committed:
refactor: ProductInfo 타입에 커스텀 이니셜라이저 구현

- 프로퍼티가 default value 를 갖는 형태가 아니라, 이니셜라이저에서 initial value 를 입력해주는 방식으로 변경
- required 가 아니고 default value 또한 필요하지 않은 값에 대해선 옵셔녈 처리
- [APIURL] 타입에서 apiHost -> openMarketAPIHost 네이밍 구체화 및 private 접근제어 부여

yeahg-dev committed:
feat: requestProductRegistration, createBody 메서드 구현

- ImageFile 타입 구현
- Data+Extension 구현

Jager-yoo committed:
feat: requestProductEdit 메서드 구현

yeahg-dev committed:
feat: requestProductSecret 메서드 구현

Jager-yoo committed:
feat: requestProductDeletion 메서드 구현

yeahg-dev committed:
feat: requestProductDetail 메서드 구현

Jager-yoo committed:
feat: requestProductsListPage 메서드 구현

yeahg-dev committed:
refactor: Parserable 프로토콜 생성 및 프로토콜 기본 구현

Jager-yoo committed:
refactor: APIRequestProtocol 프로토콜 기본구현 생성

- execute, createBody 같은 메서드를 프로토콜 기본구현으로 만들어서, 프로토콜을 채택만해도 메서드를 사용할 수 있게 함
- 그리고 APIRequestProtocol 이 Parserable 프로토콜을 상속하게 하여 encode, decode 메서드 또한 사용할 수 있게 함

Jager-yoo committed:
refactor: Parserable 프로토콜 기본구현의 decode 메서드에 dateDecodingStrategy 추가

Jager-yoo committed:
refactor: 기본 URL 네이밍 수정 openMarketAPIHost -> apiHost

kcharliek commented on OpenMarket/MockJSONParserTests/MockJSONParser.swift:
```diff
@@ -0,0 +1,25 @@
+
+enum MockJSONParser<T: Decodable> {
+    
+    static func decode(from jsonName: String) -> T? {
```
static method는 피해주세요. Apple의 Foundation Framework에는 static util이 거의 없답니다. 구조체로 만드는 방향으로 수정해주세요.

Jager-yoo replied:
decode, encode 메서드를 정의한 타입을 `구조체`로 변경했습니다!
생성 비용이 큰 `JSONEncoder()`, `JSONDecoder()`, `DateFormatter()` 는 인스턴스 프로퍼티로 생성해두고 재사용하도록 만들었습니다.

그리고 유닛 테스트에서만 사용되는 decode 메서드 또한 구조체로 가지고 들어오고 MockJSONParser.swift 파일은 삭제했습니다!

```swift
struct JSONCodable {
    
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private let dateFormatter = DateFormatter()

    func decode<T: Decodable>(from data: Data) -> T? {
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SS"
        decoder.dateDecodingStrategy = .formatted(dateFormatter)
        guard let data = try? decoder.decode(T.self, from: data) else {
            return nil
        }
        return data
    }
    
    // 유닛 테스트에 사용되는 decode 메서드
    func decode<T: Decodable>(from fileName: String) -> T? {
        guard let dataAsset = NSDataAsset(name: fileName) else {
            return nil
        }
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SS"
        decoder.dateDecodingStrategy = .formatted(dateFormatter)
        
        let decodedData = try? decoder.decode(T.self, from: dataAsset.data)
        return decodedData
    }
    
    func encode<T: Encodable>(from data: T) -> Data? {
        guard let object = try? encoder.encode(data) else {
            return nil
        }
        return object
    }
}
```

kcharliek commented on OpenMarket/OpenMarket/Model/Currency.swift:
```diff
@@ -0,0 +1,14 @@
+
+enum Currency: String, Codable {
+    
+    case krw = "KRW"
```
enum으로 잘 나눠주셨어요.

kcharliek commented on OpenMarket/OpenMarket/Model/ProductDetail.swift:
```diff
@@ -0,0 +1,36 @@
+
+struct ProductDetail: Codable {
+    
+    let id: Int
```
let으로 선언 잘 해주셨습니다. Model의 변경은 최대한 막아두어야합니다.

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequest.swift:
```diff
@@ -0,0 +1,119 @@
+    case delete = "DELETE"
+}
+
+enum APIRequest: APIRequestProtocol {
```
APIRequestProtocol의 설계 근거가 무엇인지 궁금합니다.

yeahg-dev replied:
1. 리팩터링 과정에서, 기존 APIRequest 열거형의 static function에서 공통적으로 사용되는 메서드들을 빼주기위해서 설계했습니다.
프로토콜 기본 구현으로 `execute` 와 `createBody`를 구현하면, APIRequest에서 이 프로토콜을 채택만해주면 바로 메서드를 사용할 수 있기 때문입니다.

2. 프로토콜 명이 APIRequest인 이유
`APIRequest`라는 이름으로 만들어주고 싶었으나, 이미 열거형이 동일 이름이라 임시적으로 `APIRequestProtocol`라고 네이밍했습니다. 그리고 추후 열거형에 있는 타입메서드들이 각각의 구조체로 독립될것이라 생각했습니다. 그리고 다시 프로토콜명을 `APIRequest`로 바꾸고, 구조체들이 채택하도록 구현하게 해줄 계획이었습니다.

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequest.swift:
```diff
@@ -0,0 +1,119 @@
+
+enum APIRequest: APIRequestProtocol {
+    
+    static func requestHealthChecker() {
```
static method는 지양하셔야합니다. 리팩토링 전에 우선 APIRequest를 class로 바꾸고 일반 method로 수정해주세요. 
`APIRequest.requestHealthChecker` -> `APIRequest().requestHealthChecker`

Jager-yoo replied:
### 디스코드로 나눈 대화 메모

> (예거) 찰리! 질문이 있습니다. 😄 
> 코멘트에서 APIRequest 타입을 `클래스` 로 바꾸고 각각의 request 를 func 로 수정해달라고 말씀해주셨는데요!
>
> 지난 번 찰리가 직접 써주신 예시에서는 각각의 reqeust 를 `구조체` 타입으로 구현하셨거든요. 그래서 저희도 후자로 리팩토링을 해야겠다고 생각을 하고 있었어요!
> 혹시 클래스 와 일반 메서드로 리팩토링을 말씀하신 의도가 무엇인지 알려주실 수 있을까요??? 🤔
>
> (찰리) 아 요부분 class도 좋고 구조체도 좋은데 `구조체`가 더 좋을 것 같아요. enum 사용을 하지 않을 것을 권해드린 상황으로 이해해주시면 될거같아요!

Jager-yoo replied:
위 대화에 근거하여
`APIRequest` 타입은 클래스가 아닌 `구조체`로 변경했습니다! 😄

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequest.swift:
```diff
@@ -0,0 +1,119 @@
+        images: [ImageFile],
+        _ completion: @escaping Handler
+    ) {
+        guard let url = APIURL.productRegistration.url else { return }
```
API 메서드의 구성이, 1. URL가져오기, 2. URLRequest 만들기, 3. httpMethod 담기, 4. header 담기, 5. body 담기 6. execute 로 되어있음을 확인했습니다. 
API가 추가될 때 마다 이 코드를 반복해서 작성하는 방식이 아니라, 추상화 레벨을 높혀서 일괄로 처리했으면 좋겠는데요. 이 방향으로 고민 부탁드리겠습니다.

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequestProtocol.swift:
```diff
@@ -0,0 +1,60 @@
+
+extension APIRequestProtocol {
+    
+    typealias Handler = (Result<Data, Error>) -> Void
```
요청 결과를 Data로 내려주는 것이 아니라, 각 API 호출 결과를 가지고 Model로 Decoding까지 해서 내려주면 좋겠습니다. 현재 방식으로는 사용하는 개발자가 매번 Decoding 로직을 작성해야만 하는 점이 부담스럽습니다.

Jager-yoo replied:
이 부분 릴리와 함께 고민해봤는데요!
아직 호출 결과를 어떻게 Model 로 디코딩할지 그림을 그리기가 어려워서, `STEP 2` 진행하면서 다시 리팩토링 고민해보려고 합니다!

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequestProtocol.swift:
```diff
@@ -0,0 +1,60 @@
+        return "--\(UUID().uuidString)"
+    }
+    
+    static func execute(request: URLRequest, _ completion: Handler?) {
```
execute 로직을 구현해두고, 이 메서드를 재사용하도록 한 방식 잘 하셨습니다.

kcharliek commented on OpenMarket/OpenMarket/Network/APIURL.swift:
```diff
@@ -0,0 +1,44 @@
+
+    private static let apiHost = "https://market-training.yagom-academy.kr/"
+
+    case healthChecker
```
enum으로 URL 잘 정리 해두셨습니다. associatedValue로 파라미터 전달한 것 까지 아이디어 좋습니다.

kcharliek commented on OpenMarket/OpenMarket/Network/RequestType/ImageFile.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    enum ImageType {
+        
+        case jpg
```
ImageType 정리 아~주 잘 하셨습니다 . 개발자에게 더 많은 옵션을 편리하게 제공할 수 있다는 점에서 좋아보입니다.

kcharliek commented on OpenMarket/OpenMarket/Network/APIRequestProtocol.swift:
```diff
@@ -0,0 +1,60 @@
+        dataTask.resume()
+    }
+    
+    static func createBody(params productInfo: NewProductInfo, images: [ImageFile]) -> Data? {
```
APIRequestProtocol 과 NewProductInfo 가 강하게 의존하고 있는 점이 걸립니다. 이 부분 수정되었으면 합니다.

Jager-yoo replied:
해당 메서드는 `<상품 등록>` Request 에서 새로 등록할 `상품 타입(NewProductInfo)`을 httpBody 에 인코딩해서 넣어주는 기능을 합니다!

따라서, 여기는 파라미터로 NewProductInfo 타입이 들어가야만 하는 상황입니다.
혹시 반드시 그 타입이 들어가야 할 때도 `강하게 의존`하면 안 되는 걸까요?? 🤔 

이 부분을 `Encodable 프로토콜`을 채택하고 있는 타입만 들어올 수 있게 만들어서 의존도를  낮출 수 있지만, Encodable 프로토콜을 채택한 다른 타입이 들어오게 될 가능성이 생깁니다. -> 그러면 통신에 실패하게 됩니다.

그래서 오히려 NewProductInfo 타입 하나만 들어올 수 있게 제한하는 게 좋지 않을까 생각했습니다!

kcharliek commented on OpenMarket/OpenMarket/Network/RequestType/ProductInfo.swift:
```diff
@@ -0,0 +1,52 @@
+
+import Foundation
+
+struct NewProductInfo: Encodable {
```
Encodable 이 왜 필요한 걸까요 ?

yeahg-dev replied:
request의 바디에 실어보낼 데이터 타입이여서 Encodable을 채택시켰습니다!

kcharliek commented:
### 고민한 내용에 대해

1. dateDecodingStrategy.iso8601 사용 실패

strategy에 사전 정의된 format을 사용한 경험이 오히려 적었던 것 같아요. custom data format 잘 사용해주셨습니다 :)

2. convertFromSnakeCase 전략 🆚 CodingKeys 열거형

convertFromSnakeCase 을 사용하게되면 서버가 주는 key와 내 property name에 의존성이 생깁니다. 만일 서버 key가 바뀌면 내 이름도 바꿔야할 것이고, 이를 사용하게 되는 위치의 코드도 함께 수정해야할 것입니다. 이는 서버에 의존하게 되는 것이므로 좋은 상황은 아니라고 생각합니다. CodingKeys를 사용하는 것은 서버와 클라이언트의 의존성을 끊어주는 아주 멋진 스타일이라고 생가하고 이를 잘 사용해주셨습니다.

3. Request에서 boundary를 꼭 써야할까?

`part별 name을 Content-Disposition에 넣어주고, data를 아래에 추가해주는게 맞는 방법일까요?` 라는 질문은 제가 답변해 드리기 보다, 직접 구글링 해서 찾으시는 편이 좋을 것 같아요. HTTP통신은 정해져 있는 규칙이고 우리는 규칙을 찾아서 적용하기만 하면 됩니다. 규칙을 보고 코딩이 어렵다면 stackoverflow가 알려줄 것 입니다.  [let me google that for you!](https://www.google.com/search?q=ios+swift+multipart+formdata) 그보다 `append` 가 아니라 `appendString` 을 사용한 점이 눈에 띄는데요. 이렇게 해도 동작에 문제가 없던가요 ?

kcharliek commented:
### 질문에 대해

1. 어떻게 추상화 레벨을 높일 수 있을까요?

예거가 Discord로 저에게 잘 말해주었는데요.
``` text
1. 작은 부분이 반복되더라도, 일단 구현되는 걸 만들어보자!
2. 구현을 마무리한 뒤에, 반복되는 코드를 포착해서 리팩토링해보자
3. 공통적으로 쓰이는 메서드가 있다면 프로토콜 기본구현 으로 분리시켜보자!
```
의 단계를 거치면 됩니다. 제가 예거 & 릴리와의 첫 미팅에서 구조 방향성을 바로 잡아드린 것은 위단계를 건너 뛴 것으로, 여러분이 해당 설계의 장점을 이해하지 못하셨던 것이 당연하다고 생각하고요. 이 부분에 대해 미안하게 생각합니다. 이 PR에서 1단계 잘 해주셨습니다. 이제 2, 3단계를 하시면 될텐데요. 같이 이야기 해보셔도 좋고, 제가 코멘트 드린 부분 참고하셔서 리팩토링 해주셔도 좋습니다 :) 가능하면 저에게 많이 물어봐주세요. 많이 어려운 설계인 점 저도 인지하고있습니다 ㅠㅠ 

2. boundary는 request마다 생성해주어야하는 것 인지

request마다 생성해 주어야할 것 같습니다. boundary는 request의 body안에 포함되어있는 것이니까요. request 생성 시점 마다 다시 세팅해주어야하는 것이 당연해보입니다.

yeahg-dev committed:
refactor: createBody 메서드의 request body 내용 수정

Jager-yoo committed:
refactor: APIRequest 타입을 클래스로 수정

- APIRequest 타입을 클래스로 바꾸고 각각의 request 를 전부 인스턴스 메서드로 변경
- APIRequestProtocol 프로토콜 기본구현의 타입 메서드를 전부 메서드로 변경
- 기존 Parserable 프로토콜 기본구현을 JsonCodable 이라는 네이밍의 구조체로 변경하고 타입 메서드를 인스턴스 메서드로 변경
- 유닛 테스트에서만 사용되던 MockJsonParser 를 삭제하고 JsonCodable 타입의 메서드로 가져옴

Jager-yoo committed:
refactor: JSONCodable 구조체 타입 네이밍 수정

yeahg-dev committed:
refactor: APIRequest의 타입을 클래스에서 구조체로 수정

Jager-yoo committed:
refactor: 모든 APIRequest 메서드가 구조체로 정의되고 공통의 프로토콜을 채택하여 추상화되도록 구현

- APIRequest 프로토콜 정의
- 각각의 APIRequestType 들이 APIRequest 프로토콜 채택 후 구현
- HTTPMethod 열거형 별도 파일로 분리

yeahg-dev committed:
refactor: APIExecutor 타입 생성 후 APIRequestProtocol의 메서드 이동

Jager-yoo committed:
test: execute 메서드 테스트를 위한 MockURLSession 구현

yeahg-dev committed:
test: MockURLSession 테스트 케이스 추가

Jager-yoo committed:
test: Health Check API 테스트 케이스 추가

Jager-yoo committed:
fix: GET Request 의 header 에 boundary 추가

- body 가 nil 인 경우 Content-Type 에 boundary 를 넣어주지 않았으나, 그 경우 통신이 제대로 이뤄지지 않음
- GET Request 2가지의 header 에 boundary 추가 후, viewDidLoad() 메서드 내부에 테스트 코드를 작성해서 통신이 제대로 이뤄지고 있음을 확인함

Jager-yoo commented:
@kcharliek 찰리! 안녕하세요. 😄 
스텝1 리팩토링을 마치고 다시 코멘트 드립니다! 잘 챙겨주셔서 항상 감사합니다.

### 🛠️ 리팩토링한 부분

- 1️⃣ 모든 APIRequestType 을 메서드가 아닌 각각 `구조체(struct)`로 변경했습니다.
- 2️⃣ 변경한 구조체가 공통의 프로토콜(APIRequest)을 채택하여 `API 명세`를 구현하도록 했습니다.
- 3️⃣ 유사하게 반복되는 HTTP header, body 를 만드는 로직을 `추상화`했습니다.

### 🏁 네트워크 테스트

- 저희가 열심히 만든 APIRequest 가 제대로 작동하는지 테스트 해보기 위해, `유닛 테스트`를 만들었으나, 정상적으로 테스트가 동작하지 않았습니다. execute 의 `클로저`가 호출되지 않아서 switch문을 타지 않는 버그가 있었습니다. 
- 유닛 테스트를 위한 디버깅에 시간이 너무 오래 걸릴 것 같아서, 저희는 대신 viewController 의 `viewDidLoad()` 메서드 내부에서 실제 APIRequest 를 execute 시켜서 `통신이 제대로 작동하고 있음을 검증`할 수 있었습니다.

![image](https://user-images.githubusercontent.com/71127966/148942793-348a65cc-aa6d-419e-bdf1-bb94cabaf375.png)

kcharliek commented on OpenMarket/OpenMarket/Network/APIExecutor.swift:
```diff
@@ -0,0 +1,57 @@
+
+import Foundation
+
+protocol URLSessionProtocol { }
```
URLSessionProtocol 의 설계 의도가 궁금합니다

kcharliek replied:
> URLSessionProtocol 은 Executor 에서 URLSession 과 MockURLSession 을 주입받고 사용하기 위해서 타입으로서 만든 깡통 프로토콜이었습니다

디스코드를 통해 위의 이유로 사용하신 것을 전달받았습니다. 제가 조금 헷갈리는 부분은 APIExecutor에서 주입은 받고있으나 실제로 사용하고 있지 않더라고요. 이 부분에 대해서 왜 그렇게 설계되어있는지, 향후 개선을 한다면 어떤식으로 생각하고 계신지 궁금합니다 ~!

yeahg-dev committed:
test: MockURLSession관련 타입과 테스트케이스 삭제
